### DISCUSSION TITLE
Why OpenTelemetry enabled by default?

### DISCUSSION THREAD

**Discussion by ppatierno** [quarkusio/quarkus#31635] [Q&A] [answered]
May I ask why the property `quarkus.opentelemetry.enabled` is `true` by default so having the OpenTelemetry tracer being instantiated by default?
Tracing is something optional so I would expect it to be disabled by default, not the other way around.

[a comment by an automated account is not shown]

**Comment by radcortez** [comment, marked as answer]
The configuration is only applied when the Quarkus OpenTelemetry Extension is added to a Quarkus project. We assume that if the user added that dependency, OpenTelemetry is required (equivalent to enabling a configuration).

On the other hand, requiring the user to add the dependency, plus enable a flag to have OpenTelemetry working, was counter-intuitive, but I guess that some users would prefer otherwise :)

**Reply by ppatierno** [reply]
I am not sure I follow. In our project (still the Strimzi HTTP Kafka bridge ;-)) we provide the user the possibility to have tracing enabled. So we need to add the opentelemetry extension upfront because we are providing the application.
Then via configuration the user can enable or disable the tracing, and my expectation as a user would be that it is disabled by default (as it is today for our application without Quarkus).
I think you should considering third-party software as we provide, not just when the user is writing applications.

**Reply by radcortez** [reply]
Usually, that is how Quarkus operates with extensions (you add them, and they enable something, with configurations to disable certain pieces). If you think of extensions like RESTEasy, Bean Validation, OpenAPI, etc., you don't need to enable them explicitly. They become enabled once they are added to the project. We are trying to follow the same philosophy here so each extension has consistent behaviour across the ecosystem.

Unfortunately, changing the default now would be a significant breaking change :(

You can still disable telemetry within the third party and control the toggle with a specific third-party configuration.

**Reply by brunobat** [reply]
These are build time properties. You cannot override them at runtime. 
As Roberto is mentioning, If you have the extension as a dependency you are already stating that want the functionality. 
On a future version (in the works) we will provide the ability to disable or enable all OpenTelemetry at runtime, though.

**Reply by ppatierno** [reply]
> These are build time properties. You cannot override them at runtime.

Why not? When I said at runtime I meant that the user can provide an application.properties out of the resources so from an external path and he can disable or enable OpenTelemetry as he wants.
This is what I have right now, and what I ported to Quarkus. This is how our application works; you have to provide your own config file.
Tbh if I could agree with RESTeasy extension, because I am expressing the intention of using it for developing my REST interface, I see differently for something like tracing. You want to trace or not, depending on the use case, maybe just sometimes for testing purposes or getting insights on performance but then disable forever. It sounds counterintuitive having tracing enabled by default imho.
Btw I can understand it's not possible to change the default easily right now (maybe in future?) but not sure that following the same approach across all extensions makes really sense.
For now, I will go disabling it via the config property by default allowing the user to re-enable it if needed.

**Reply by brunobat** [reply]
> Why not?

Build time properties only work when you are building the project.

**Reply by maxandersen** [reply]
The user would need to rebuild or rather reaugment the app to enable it Afaics. 

The advantage of the buildtime setting is you can avoid the overhead when you don't need it. 
It's part of what makes it possible to gain speed/memory utilization. 

Agree that there should be global on/off for telemetry at runtime as @brunobat. But that would be something else than the buildtime control.

**Reply by ppatierno** [reply]
So at this point what am I doing wrong?
I have the dependency in my project but I provide an application.properties file on the command line (not from the resources folder) with the quarkus.opentelemetry.enabled property set to false. I am using -Dquarkus.config.locations to pass the path of the configuration file.

This way, I have two results: a Tracer instance, that I use in my application, is not injected (I am using an Instance<Tracer> and it results to be not resolvable) as well as the HTTP requests don't generate any traces. So it looks like tracing is actually disabled.
If I change the property to true and restart the application, it is enabled (Tracer instance is injected and the HTTP requests are producing traces out of box)

**Reply by brunobat** [reply]
Are you using dev mode, starting a JVM server or a native image server?

**Reply by ppatierno** [reply]
I am in dev mode right now

**Reply by brunobat** [reply]
If you build each time you start devmode it will always looks like the build time property is runtime. Not quite.
Try to start the server with `java -jar ...` with a changed build time property value. You will see no behaviour changes.

**Reply by ppatierno** [reply]
I am usually running the app this way ...

```shell
mvn quarkus:dev -Dquarkus.args="--config-file config/application.properties"
```

Where the configuration is in an external application.properties file, where I have my own config params and quarkus OpenTelemetry related. Something like this.

```shell
#Bridge related settings
bridge.id=my-bridge
# uncomment one the following lines (bridge.tracing) to enable Jaeger tracing, check the documentation how to configure the tracer
# OpenTracing support
#bridge.tracing=jaeger
# OpenTelemetry support
#bridge.tracing=opentelemetry

#Apache Kafka common
kafka.bootstrap.servers=localhost:9092

#Apache Kafka producer
kafka.producer.acks=1

#Apache Kafka consumer
kafka.consumer.auto.offset.reset=earliest

#HTTP related settings
http.host=0.0.0.0
http.port=8080
#Enable CORS
http.cors.enabled=false
http.cors.allowedOrigins=*
http.cors.allowedMethods=GET,POST,PUT,DELETE,OPTIONS,PATCH

#Enable consumer
http.consumer.enabled=true

#Enable producer
http.producer.enabled=true

#Quarkus related configuration

#OpenTelemetry support
quarkus.application.name=${OTEL_SERVICE_NAME}
quarkus.opentelemetry.enabled=true
quarkus.opentelemetry.tracer.exporter.otlp.endpoint=${OTEL_EXPORTER_OTLP_ENDPOINT}
```

I also have a SmallRye config interceptor to "translate" things like `http.port` to `quarkus.http.port` and so on.

**Reply by ppatierno** [reply]
Thanks to @brunobat we had an offline call and not things are clearer. I need to run the application as jar and see that it won't work for me :-)

**Reply by ppatierno** [reply]
Thanks to @brunobat we had an offline call and not things are clearer. I need to run the application as jar and see that it won't work for me :-)
